cmd/frontrunning/main.go

```diff
@@ -0,0 +1,198 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"log"
+	"math"
+	"math/big"
+	"net/http"
+	"time"
+
+	"github.com/Amenokal-Labs/mev-mantis.git/pkg/utils"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/gethclient"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+func main() {
+	ethclient, err := ethclient.Dial("https://goerli.infura.io/v3/" + utils.GetKey("INFURA_KEY"))
+	if err != nil {
+		log.Fatal("[0] ", err)
+
+	}
+
+	rpc, err := rpc.Dial("wss://goerli.infura.io/ws/v3/" + utils.GetKey("INFURA_KEY"))
+	if err != nil {
+		log.Fatal("[1] ", err)
+	}
+	subscriber := gethclient.New(rpc)
+	fmt.Println("client connected..")
+
+	// txs := make(chan *types.Transaction)
+	// _, err = rpcClient.SubscribeFullPendingTransactions(context.Background(), txs)
+	// if err != nil {
+	// 	panic(err)
+	// }
+
+	for {
+		pc, _ := ethclient.PendingTransactionCount(context.Background())
+		fmt.Println("\nPending count:", pc)
+
+		txn, marshalledTxn, from := getTxn(ethclient, subscriber)
+		calldata, contract := getCalldata(txn, marshalledTxn)
+		if contract == "0x" {
+			continue
+		}
+		fmt.Println("    Tx Calldata:", calldata)
+		fmt.Println("     Tx address:", txn.To())
+		fmt.Println("\nContract code:", contract[:26], "...")
+
+		address := from.String()
+		if replaceAddress(calldata, address) != calldata {
+			sendTxn(ethclient, txn)
+		}
+
+		fmt.Println("___________________________")
+		time.Sleep(4 * time.Second)
+	}
+}
+
+func getTxn(_ethclient *ethclient.Client, _subscriber *gethclient.Client) (txn *types.Transaction, marshalledTxn []byte, from common.Address) {
+	hashes := make(chan common.Hash)
```

Comment:
what does this line do?

Reply:
The very next line is:
`_, err := _subscriber.SubscribePendingTransactions(context.Background(), hashes)`

`SubscribePendingTransactions` subscribes to new pending transaction hashes, which `hashes` as a second parameter of type `chan<- common.Hash`

The line u mentioned initializes that parameter.

Reply:
ok, I understand this :)

but I didn't know the meaning of "make", "chan", etc. I've just read a bit about channel and go-routine to understand

Reply:
I apologize if my response seemed like overexplaining. As an AI language model, I am designed to provide clear and concise answers to questions... lmao